### PR TITLE
Support vector maps in google module

### PR DESCRIPTION
**Pull request by felixpalmer**
Implementation of https://github.com/visgl/deck.gl/issues/5980.

I still need to add tests and examples, but I would like to get some feedback on the general approach taken before starting on this.

I would also appreciate any pointers on how to do deal with recovering from a WebGL context loss/restore

### CODE REVIEW

**Comment by felixpalmer** [comment]
The examples in `app2.js` & `app3.js` are more as an aid for development/debugging and should not be reviewed

**Comment by felixpalmer** [comment]
Changes made:

- Switch base map to vector
- Migrate to new WebglOverlayView
- Using Mapbox overlay as an reference, move the rendering of the deck.gl layers into the gl context of Google maps
- Use matrices from Google (view and projection) to correctly set projection with z-buffer matched
- Automatically fallback to old overlay if using a raster map (this can only be determined at runtime as it is configured in the Google Cloud Console)
- Confirm tilt and rotation of camera works
- Handle mouse events (click, doubleclick, mouse move & mouse out)
- Support interleaving (e.g. between Google buildings and deck markers)
- Make sure animation triggered in Google maps works (deck.gl layers are also animated)

File: modules/google-maps/src/utils.js [hunk 1]
```diff
@@ -145,8 +150,75 @@ export function getViewState(map, overlay) {
...
+  });
+  const focalDistance = 0.5 * projectionMatrix[5];
+
+  return {
+    width,
+    height,
+    views: [
+      new MapView({
+        id: 'google-maps-overlay-view',
+        projectionMatrix
```
Comment (Pessimistress) [inline modules/google-maps/src/utils.js:196]: Just place projectionMatrix in viewState?

File: examples/get-started/pure-js/google-maps/app2.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
+/* global document, google, window */
```
Comment (Pessimistress) [inline examples/get-started/pure-js/google-maps/app2.js:1]: The get-started examples are intended to be boilerplates, not showcases of library capabilities. Let's keep the dataset aligned with all other get-started examples. For dev testing purpose, you can add apps to test/apps. We will make a full standalone demo of context interleaving by updating https://deck.gl/examples/google-maps/

Comment (alasarr) [reply examples/get-started/pure-js/google-maps/app2.js:1]: Agree. I think we can remove the examples in this PR. Once we land this, we can work on a cool 3D example to replace https://deck.gl/examples/google-maps

File: modules/google-maps/src/utils.js (hunk 1 above)
Comment (felixpalmer) [reply modules/google-maps/src/utils.js:194]: That doesn't work as the default null projectionMatrix from the `MapView` overrides it in https://github.com/visgl/deck.gl/blob/2601f2869ffca4c48c18c880582631ac3435e5ca/modules/core/src/views/view.js#L154

File: examples/get-started/pure-js/google-maps/app2.js (hunk 2 above)
Comment (felixpalmer) [reply examples/get-started/pure-js/google-maps/app2.js]: I've removed the examples from the PR

File: examples/get-started/pure-js/google-maps/app.js [hunk 3]
```diff
@@ -6,9 +6,11 @@ import {GeoJsonLayer, ArcLayer} from '@deck.gl/layers';
 const AIR_PORTS =
   'https://d2ad6b4ur7yvpq.cloudfront.net/naturalearth-3.3.0/ne_10m_airports.geojson';
 
+const GOOGLE_MAP_ID = 'f865292e93a85c6c';
```
Comment (Pessimistress) [inline examples/get-started/pure-js/google-maps/app.js:9]: What is this magic string? Is it something that all users would want to use?

File: modules/google-maps/src/google-maps-overlay.js [hunk 4]
```diff
@@ -66,18 +73,56 @@ export default class GoogleMapsOverlay {
...
+    if (isVectorMap) {
+      overlay.onAdd = () => {};
+      overlay.onContextLost = this._onContextLost.bind(this);
+      overlay.onContextRestored = this._onContextRestored.bind(this);
+      overlay.onDraw = this._onDrawVector.bind(this);
+    } else {
+      overlay.onAdd = this._onAdd.bind(this);
+      overlay.draw = this._onDrawRaster.bind(this);
+    }
+    overlay.onRemove = this._onRemove.bind(this);
```
Comment (Pessimistress) [inline modules/google-maps/src/google-maps-overlay.js:96]: Can we define these methods in ES6 syntax? E.g.

```js
const {VECTOR, UNINITIALIZED} = google.maps.RenderingType;

class GoogleMapsOverlay {
  _createOverlay(map) {
    this.renderingType = map.getRenderingType();
  }

  onAdd() {
    if (this.renderingType !== VECTOR) {
      this._deck = createDeckInstance(...);
    }
  }

  draw() {
    // old API - draw raster
  }

  onDraw() {
    // new API - draw vector
  }

  onRemove() {
  }
}
```

File: modules/google-maps/src/utils.js [hunk 5]
```diff
@@ -29,7 +31,7 @@ export function createDeckInstance(map, overlay, deck, props) {
   deck = new Deck({
     ...props,
     style: null,
-    parent: getContainer(overlay, props.style),
+    parent: overlay && getContainer(overlay, props.style),
```
Comment (Pessimistress) [inline modules/google-maps/src/utils.js:34]: When would `overlay` be null?

File: modules/google-maps/src/utils.js [hunk 6]
```diff
@@ -57,7 +59,14 @@ function getContainer(overlay, style) {
   const container = document.createElement('div');
   container.style.position = 'absolute';
   Object.assign(container.style, style);
-  overlay.getPanes().overlayLayer.appendChild(container);
+  if (overlay.getPanes) {
```
Comment (Pessimistress) [inline modules/google-maps/src/utils.js:62]: Add comment on old/new code path

File: modules/google-maps/src/utils.js (hunk 1 above)
Comment (Pessimistress) [reply modules/google-maps/src/utils.js:194]: Shall we just remove the `null` fallbacks in View constructor? They don't seem to provide any value.

File: examples/get-started/pure-js/google-maps/app.js (hunk 3 above)
Comment (felixpalmer) [reply examples/get-started/pure-js/google-maps/app.js:9]: It is defined on a per-user basis, they obtain from the Google cloud console. I've removed it an instead it is now provided via an environment variable, like we do with `GOOGLE_MAPS_API_KEY`.

File: modules/google-maps/src/utils.js (hunk 5 above)
Comment (felixpalmer) [reply modules/google-maps/src/utils.js:34]: This is a mistake, I've removed it

File: examples/get-started/pure-js/google-maps/app.js (hunk 3 above)
Comment (Pessimistress) [reply examples/get-started/pure-js/google-maps/app.js:9]: Is it just for styling? Is it possible to create a map without it?

Comment (felixpalmer) [reply examples/get-started/pure-js/google-maps/app.js]: It is required to create a vector map, not just for styling. I have updated the README

File: examples/get-started/pure-js/google-maps/README.md [hunk 7]
```diff
@@ -9,13 +9,14 @@ with [webpack-dev-server](https://webpack.js.org/guides/development/#webpack-dev
 
 ## Usage
 
-To run this example, you need a [Google Maps API key](https://developers.google.com/maps/documentation/javascript/get-api-key). You can either set an environment variable:
+To run this example, you need a [Google Maps API key](https://developers.google.com/maps/documentation/javascript/get-api-key) and a map id. You can either set an environment variable:
```
Comment (Pessimistress) [inline examples/get-started/pure-js/google-maps/README.md:12]: Can you link to the map id documentation

File: modules/google-maps/src/google-maps-overlay.js (hunk 4 above)
Comment (felixpalmer) [reply modules/google-maps/src/google-maps-overlay.js:96]: Sorry, I don't follow here. Are the methods `_onAdd`, `_onContextRestored` etc not already using the correct syntax? We need to bind them to the `google.maps.WebglOverlayView` object as per Google's docs

**Review by Pessimistress — APPROVED** [review]
Can be a follow up PR - the module documentation should be updated to explain the compatibility/limitations with both Google Maps versions.

File: modules/google-maps/src/utils.js (hunk 1 above)
Comment (felixpalmer) [reply modules/google-maps/src/utils.js:197]: I agree, the defaults seem to get in the way. I've remove the `projectionMatrix` and `viewMatrix` defaults and updated the overlay to pass the `projectionMatrix` directly as part of the `viewState`

File: examples/get-started/pure-js/google-maps/README.md (hunk 7 above)
Comment (felixpalmer) [reply examples/get-started/pure-js/google-maps/README.md:12]: Done

**Pull request merged**